lib/parsers/pdf_parser.rb

```diff
@@ -14,62 +15,58 @@ class FormatParser::PDFParser
   # this. The only way of solving this correctly is by adding
   # different types of PDF's in the specs.
   #
-  COUNT_MARKERS = ['Count ']
-  EOF_MARKER    = '%EOF'
+  EOF_MARKER = '%%EOF'
 
   def call(io)
     io = FormatParser::IOConstraint.new(io)
 
     return unless safe_read(io, 9) =~ PDF_MARKER
 
-    attributes = scan_for_attributes(io)
+    @xref_table = FormatParser::PDFParser::XRefTable.parse(io)
 
     FormatParser::Document.new(
       format: :pdf,
-      page_count: attributes[:page_count]
+      page_count: page_count(io)
     )
   end
 
   private
 
-  # Read ahead bytes until one of % or / is reached.
-  # A header in a PDF always starts with a /
-  # The % is to detect the EOF
-  #
-  def scan_for_attributes(io)
-    result = {}
-
-    while read = safe_read(io, 1)
-      case read
-      when '%'
-        break if safe_read(io, EOF_MARKER.size) == EOF_MARKER
-      when '/'
-        find_page_count(io, result)
-      end
-    end
+  def page_count(io)
+    page_count = nil
 
-    result
-  end
+    @xref_table.each do |xref|
+      io.seek(xref.offset)
+
+      page_count = case safe_read(io, 32)
+                   when /\/Type\s*\/Catalog/
+                     read_header(io, xref, /\/Pages (\d+)/)
```

Comment:
This one is probably not a very good idea. The `/Pages` element usually contains a reference value (a ref) which is expressed as `<object id> <object generation value> R`. With this regexp you will very likely capture the object ID which has nothing to do with the page count in the document. So you either need to look for something that is guaranteed to contain a literal integer for pagecount (like the `/Count` name) or you need to first figure out what each object from the xref table contains, and _then_ be able to parse (and resolve!) those references. Look at this presentation for example - https://www.adobe.com/technology/pdfs/presentations/KingPDFTutorial.pdf In the "Hello world" even the pages dictionary contains a reference value under `/Pages` name.

In other words, if we do not want to really tokenize and parse the PDF objects with all the related types we got to be super-extra-careful with applying regexes. I'm sorry, I didn't invent this file format 🤷‍♂️

Reply:
That being said, you _can_ use this to find the referenced object in your xref table, seek to it and read the actual '/Pages` object - then you can quickly find the `/Count` which lives inside the `/Pages`. That's what the reference points to.

Reply:
Oh yes. Forgot all about that `/Pages` could also be a reference... 🙄 Yes, than this will break h orribly. Funny that my specs succeed though. 🤔 Will add some more for this specific case though.

Reply:
You also want to extend your regexp to include one non-digit for the possible subsequent chars, or at least for a subsequent whitespace or dictionary close (`>>`) to ensure your read didn't truncate the name and the value.